optuna/samplers/_cmaes.py

```diff
@@ -369,10 +380,46 @@ def sample_relative(
             search_space, transform_step=not self._with_margin, transform_0_1=True
         )
 
-        optimizer, n_restarts = self._restore_optimizer(completed_trials)
+        if self._initial_popsize is None:
+            self._initial_popsize = 4 + math.floor(3 * math.log(len(trans.bounds)))
+
+        store_optimizer = False
+        optimizer = self._restore_optimizer(completed_trials)
+        optimizer_inited = False
         if optimizer is None:
-            n_restarts = 0
-            optimizer = self._init_optimizer(trans, study.direction, population_size=self._popsize)
+            optimizer = self._init_optimizer(
+                trans, study.direction, population_size=self._initial_popsize
+            )
+            optimizer_inited = True
```

Comment:
I am currently looking for a way to make this logic simpler. If we need to reinitialize the optimizer for different options, how about having a different system_attr to store the optimizer for each option?

Specifically, there is a way to store the optimizer in the key obtained by the following function.

```diff
diff --git a/optuna/samplers/_cmaes.py b/optuna/samplers/_cmaes.py
index a8731773f..6f1949465 100644
--- a/optuna/samplers/_cmaes.py
+++ b/optuna/samplers/_cmaes.py
@@ -46,7 +46,7 @@ _SYSTEM_ATTR_MAX_LENGTH = 2045


 class _CmaEsAttrKeys(NamedTuple):
-    optimizer: str
+    optimizer: Callable[[int], str]
     generation: Callable[[int], str]
     popsize: Callable[[], str]
     n_restarts: Callable[[], str]
@@ -534,6 +534,9 @@ class CmaEsSampler(BaseSampler):
         else:
             attr_prefix = "cma:"

+        def optimizer_key_template(restart: int) -> str:
+            return attr_prefix + f"{self._restart_strategy}:n_restart_{restart}:optimizer"
+
         def generation_attr_key_template(restart: int) -> str:
             if self._restart_strategy is None:
                 return attr_prefix + "generation"
@@ -555,7 +558,7 @@ class CmaEsSampler(BaseSampler):
                 return attr_prefix + "{}:n_restarts".format(self._restart_strategy)

         return _CmaEsAttrKeys(
-            attr_prefix + "optimizer",
+            optimizer_key_template,
             generation_attr_key_template,
             popsize_attr_key_template,
             n_restarts_attr_key_template,
```

What do you think?

Reply:
Ok, I will try this approach.

Reply:
I updated code by applying this approach and it looks works well. Could you take another look?

Reply:
@c-bata I updated code by applying this approach and it looks works well. Could you take another look?